File: db4o.net/Db4objects.Db4o.CS.Optional/native/Db4objects.Db4o.CS/Monitoring/MonitoredSocket4Base.cs
/* Copyright (C) 2009   Versant Inc.   http://www.db4o.com */
#if !CF && !SILVERLIGHT

using Db4objects.Db4o.CS.Foundation;

namespace Db4objects.Db4o.CS.Monitoring
{
	public abstract class MonitoredSocket4Base : Socket4Decorator
	{
		protected MonitoredSocket4Base(ISocket4 socket) : base(socket)
		{
		}

		public override void Write(byte[] bytes, int offset, int count)
		{
			base.Write(bytes, offset, count);
			
			Counters().BytesSent().IncrementBy(count);
			Counters().MessagesSent().Increment();
		}

		public override int Read(byte[] buffer, int offset, int count)
		{
			int bytesReceived = base.Read(buffer, offset, count);
			Counters().BytesReceived().IncrementBy(bytesReceived);

			return bytesReceived;
		}

		protected abstract NetworkingCounters Counters();
	}
}

#endif

File: db4o.cs.optional/src/com/db4o/cs/monitoring/MonitoredSocket4Base.java
/* Copyright (C) 2009 Versant Inc. http://www.db4o.com */

package com.db4o.cs.monitoring;

import java.io.*;

import com.db4o.cs.foundation.*;

import decaf.*;

/**
 * @exclude
 */
@decaf.Ignore(unlessCompatible=Platform.JMX)
abstract class MonitoredSocket4Base extends Socket4Decorator {
	public MonitoredSocket4Base(Socket4 socket) {
		super(socket);
	}

	public void write(byte[] bytes, int offset, int count) throws IOException {
		bean().notifyWrite(count);
		super.write(bytes, offset, count);
	}
	
	@Override
	public int read(byte[] buffer, int offset, int count) throws IOException {
		int bytesReceived = super.read(buffer, offset, count);
		bean().notifyRead(bytesReceived);
		
		return bytesReceived;
	}
	
	protected abstract Networking bean();
}